### ISSUE TITLE
rmcp tools default timeout is greater than scrape timeout

### ISSUE THREAD

**Issue by dannyk81** [soundcloud/ipmi_exporter#20]
Really awesome tool! thank you for sharing :smile: 

We've been hitting a small issue when using it for remote metrics collection (rmcp), it seems that the default timeout of the ipmi cli tools is quite long and in some cases we end up with scrape context deadlines so these targets are missing their respective `ipmi_up` metrics.

We can still rely on `up` to detect that the target is down but using `ipmi_up` seems more relevant in this case as it can also indicate which collector is failing/etc.

The underlying ipmi tools all have a session timeout parameter:

```
--session-timeout=MILLISECONDS
                             Specify the session timeout in milliseconds.
```

I think it would be great if we could control the timeout via the configuration (ipmi.yaml) so we can tune it to be <`scrape_timeout`

wdyt?

/cc @bitfehler

**Comment by SuperQ** [comment]
The Prometheus protocol allows you to get the scrape timeout from the server via a http header. We use this in the [blackbox_exporter](https://github.com/prometheus/blackbox_exporter/blob/bba7ef76193948a333a5868a1ab38b864f7d968a/main.go#L74-L86).

It should be easy enough to do the same here.

**Comment by SuperQ** [comment]
I guess you would need to track the execution time remaining at each command and lower the timeout as it executes the individual commands.

If it's possible, you could also use a [`sync.WaitGroup()`](https://golang.org/pkg/sync/#WaitGroup) or [`errgroup`](https://godoc.org/golang.org/x/sync/errgroup) to execute the commands in parallel. Although I don't know how many IPMI devices would fail if you tried to parallelize them.

**Comment by dannyk81** [comment]
running multiple command in parallel usually leads to issues, it is rather touchy.

I suppose we could either divide the allotted time equally between the three commands, or use your suggested approach which I find very interesting.

**Comment by bitfehler** [comment]
So with the merge of #17, you can now easily scrape as three different modules, meaning the prometheus timeout can be adjusted to match the specific tool's timeout. However, you probably will end up scraping things in parallel.
I think another sensible thing to do, now that we have a nicely modularized confing infrastructure, is exposing the timeout setting, so it can be configured. Will look into that.

**Comment by bitfehler** [comment]
Sry, to clarify maybe: create three different module, each with only a single collector.

**Comment by dannyk81** [comment]
Thanks @bitfehler, the config changes are really cool.

But scraping the same target in parallel for different collectors could indeed lead to issues, we had several cases of multiple systems querying servers using ipmi in parallel and that lead to a bunch of issues (like BMC busy and others).

So it would be great to be able to set the timeout in the exporter config or use the method suggested by @SuperQ.

In any case, thanks for looking into this :+1:

**Comment by dannyk81** [comment]
fixed by #22
